File: backend/src/main/java/com/dev/backend/resp/ResponseData.java
package com.dev.backend.resp;

import java.time.LocalDateTime;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ResponseData<T> {
    private boolean success;
    private String message;
    /** Luôn có trong JSON kể cả khi null (client thống nhất với contract API). */
    @JsonInclude(JsonInclude.Include.ALWAYS)
    private T data;
    
    private Integer code;
    private String error;
    private String path;

    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    @Builder.Default
    private LocalDateTime timestamp = LocalDateTime.now();

}

File: frontend/src/types/response-data.ts
export type ResponseData<T = unknown> =
  | {
      success: true;
      message: string;
      data: T | null;
      timestamp: string;
    }
  | ResponseDataError;

/** Lỗi từ API: `data` là map lỗi theo field (validation) hoặc bỏ qua. */
export type ResponseDataError = {
  success: false;
  message: string;
  data?: Record<string, string> | null;
  code?: number;
  error?: string;
  path?: string;
  timestamp: string;
};

export type ResponseDataSuccess<T = unknown> = Extract<
  ResponseData<T>,
  { success: true }
>;

/** Cùng pattern `yyyy-MM-dd HH:mm:ss` như `@JsonFormat` trên backend. */
export function formatResponseTimestamp(date = new Date()): string {
  const p = (n: number) => String(n).padStart(2, "0");
  return `${date.getFullYear()}-${p(date.getMonth() + 1)}-${p(date.getDate())} ${p(date.getHours())}:${p(date.getMinutes())}:${p(date.getSeconds())}`;
}

/** Lỗi sinh trên client: luôn có `timestamp`, khớp `ResponseDataError`. */
export function clientResponseError(
  message: string,
  partial?: Partial<Omit<ResponseDataError, "success" | "message">>,
): ResponseDataError {
  return {
    success: false,
    message,
    ...partial,
    timestamp: partial?.timestamp ?? formatResponseTimestamp(),
  };
}
